lib/MonadicRewrite.thy

```diff
@@ -54,34 +133,87 @@ lemma monadic_rewrite_bind:
-
 lemma monadic_rewrite_bindE:
-  "\<lbrakk> monadic_rewrite F E P f g; \<And>x. monadic_rewrite F E (Q x) (h x) (j x);
-           \<lbrace>R\<rbrace> g \<lbrace>Q\<rbrace>,- \<rbrakk>
-        \<Longrightarrow> monadic_rewrite F E (P and R) (f >>=E (\<lambda>x. h x)) (g >>=E (\<lambda>x. j x))"
+  "\<lbrakk> monadic_rewrite F E P f g; \<And>x. monadic_rewrite F E (Q x) (h x) (j x); \<lbrace>R\<rbrace> g \<lbrace>Q\<rbrace>,- \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (P and R) (f >>=E (\<lambda>x. h x)) (g >>=E (\<lambda>x. j x))"
   apply (simp add: bindE_def)
   apply (erule monadic_rewrite_bind)
-   defer
+   prefer 2
    apply (simp add: validE_R_def validE_def)
-  apply (case_tac x, simp_all add: lift_def monadic_rewrite_refl)
+  apply (case_tac x; simp add: lift_def monadic_rewrite_refl)
+  done
+
+lemmas monadic_rewrite_bind_tail
+  = monadic_rewrite_bind[OF monadic_rewrite_refl, simplified pred_and_true_var]
+
+lemmas monadic_rewrite_bind_head
+  = monadic_rewrite_bind [OF _ monadic_rewrite_refl hoare_vcg_prop, simplified pred_and_true]
+
+lemmas monadic_rewrite_bind_alt
+  = monadic_rewrite_trans[OF monadic_rewrite_bind_tail monadic_rewrite_bind_head, rotated -1]
+
+lemmas monadic_rewrite_bindE_head
+    = monadic_rewrite_bindE[OF _ monadic_rewrite_refl hoare_vcg_propE_R]
+
+lemma monadic_rewrite_bind2:
+  "\<lbrakk> monadic_rewrite F E P f g; \<And>x. monadic_rewrite F E (Q x) (h x) (j x); \<lbrace>R\<rbrace> f \<lbrace>Q\<rbrace> \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E (P and R) (f >>= (\<lambda>x. h x)) (g >>= (\<lambda>x. j x))"
+  apply (rule monadic_rewrite_imp)
+   apply (rule monadic_rewrite_trans)
+    apply (erule(1) monadic_rewrite_bind_tail)
+   apply (erule monadic_rewrite_bind_head)
+  apply simp
+  done
+
+lemma monadic_rewrite_named_bindE:
+  "\<lbrakk> monadic_rewrite F E ((=) s) f f';
+    \<And>rv s'. (Inr rv, s') \<in> fst (f' s) \<Longrightarrow> monadic_rewrite F E ((=) s') (g rv) (g' rv) \<rbrakk>
+   \<Longrightarrow> monadic_rewrite F E ((=) s) (f >>=E (\<lambda>rv. g rv)) (f' >>=E g')"
+  apply (rule monadic_rewrite_imp)
+   apply (erule_tac R="(=) s" and Q="\<lambda>rv s'. (Inr rv, s') \<in> fst (f' s)" in monadic_rewrite_bindE)
+    apply (rule monadic_rewrite_name_pre)
+    apply clarsimp
+   apply (clarsimp simp add: validE_R_def validE_def valid_def
+                      split: sum.split)
+  apply simp
   done
 
+(* FIXME: poorly named, super-specific (could do this with maybe one bind?), used in Ipc_C *)
+lemma monadic_rewrite_do_flip:
+  "monadic_rewrite E F P (do c \<leftarrow> j; a \<leftarrow> f; b \<leftarrow> g c; return (a, c) od)
+                         (do c \<leftarrow> j; b \<leftarrow> g c; a \<leftarrow> f; return (a, c) od)
+   \<Longrightarrow> monadic_rewrite E F P (do c \<leftarrow> j; a \<leftarrow> f; b \<leftarrow> g c; h a c od)
+                            (do c \<leftarrow> j; b \<leftarrow> g c; a \<leftarrow> f; h a c od)"
```

Comment:
With this being so specific, would it make sense for it to stay in `crefine`? I guess the other goal is to minimise duplication, with the only arch-generic place to then put it being `Move_C` which doesn't make sense if we don't actually want to move it.